### DISCUSSION TITLE
blog/database-normalization-3-2nf-3nf-bcnf

### DISCUSSION THREAD

**Discussion by giscus** [korECM/tw-blog#19] [Comments]
# blog/database-normalization-3-2nf-3nf-bcnf

데이터베이스 정규화 이론 중 2NF, 3NF, BCNF에 대해서 알아보자

https://www.korecmblog.com/blog/database-normalization-3-2nf-3nf-bcnf

**Comment by crossedfinger-j** [comment]
안녕하세요, 공부하는 도중 좋은 글을 발견해서 도움받으며 공부하는 학생입니다. 먼저 좋은 자료 감사합니다. 
BCNF 관련 질문이 있는데요. '그렇다면 분야가 더 자세한 정보니까 {학번, 분야}를 기본키로 설정하면 어떻게 될까? 그렇게 하면 키가 아닌 속성에서 후보키의 진부분집합에 대한 함수 종속성은 사라진다.' 이 부분에서 질문이 있습니다. 이러면 결국 '분야'->'학부'이므로 진부분 집합에 대한 함수 종속성은 아직 존재하는것 아닌가요?

**Reply by korECM** [reply]
안녕하세요. 오래 전에 작성했던 글이라 어색한 부분도 있고 오타도 있는 글인데요. 그러다 보니 헷갈리는 부분이 있었을 거라 생각합니다. 특히 외국 책에서 가져온 내용이다 보니 분야나 학부라는 개념이 더 헷갈리게 만드는 것 같네요.

그래서 BCNF 예제에서 나온 릴레이션의 상황을 다음과 같이 생각해볼게요.
- 학번 : A, 학부 : B, 분야 : C
- { A, B } 가 후보키
- B 내에 여러 개의 C가 존재. 즉, C가 더 자세한 정보이며 C가 결정되면 B도 결정됨

위 릴레이션에는 다음과 같은 자명하지 않은 함수 종속성이 존재합니다.
- {A} -> C
- {C} -> B

위 글에서 문제 삼고 있는 함수 종속성은 `{C} -> B` 입니다. 키가 아닌 속성으로부터 후보키의 진부분집합으로의 함수 종속성이 존재하니까요.

만약 {A, C}를 후보키로 바꾸면 어떻게 될까요?
- {A} -> C
- {C} -> B
후보키를 바꾼다고 함수 종속성이 바뀌진 않습니다. 각 속성이 키에 속해있냐, 그렇지 않냐가 달라집니다. 아까와 다르게 키가 아닌 속성으로부터 후보키의 진부분집합으로의 함수 종속성이 사라진 것을 알 수 있습니다. 대신, `{C} -> B`라는 후보키의 진부분집합에서 키가 아닌 속성에 함수 종속성이 생긴 것을 알 수 있습니다. 사실 `{A} -> C`라는 관점에서 후보키도 성립하지 않죠.

[학번, 학부, 분야]라는 릴레이션에서 {학번, 분야}를 기본키로 설정했을 때 키가 아닌 속성에서 후보키의 진부분집합에 대한 함수 종속성이 존재하려면 `{학부} -> 분야` 혹은 `{학부} -> 학번`의 함수 종속성이 존재해야 실제로는 존재하지 않습니다.

어떻게 설명하면 좋을까 싶어 여러 방법으로 적어보았는데 도움이 되면 좋겠습니다! 감사합니다
